daphne/src/dap_error.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2023 Cloudflare, Inc. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+
+use std::fmt::Display;
+
+use crate::{aborts::DapAbort, messages::TransitionFailure, vdaf::VdafError};
+
+/// DAP errors.
+#[derive(Debug, thiserror::Error)]
+pub enum DapError {
+    /// Fatal error. If this triggers an abort, then treat this as an internal error.
+    ///
+    /// To create an instance of this variant the [`fatal_error`] must be used. This ensures that
+    /// all fatal errors are logged with tracing when created.
+    #[error("fatal error: {0}")]
+    Fatal(#[from] FatalDapError),
+
+    /// Error triggered by peer, resulting in an abort.
+    #[error("abort: {0}")]
+    Abort(#[from] DapAbort),
+
+    /// Transition failure. This error blocks processing of a paritcular report and may, under
+    /// certain conditions, trigger an abort.
+    #[error("transition error: {0}")]
+    Transition(#[from] TransitionFailure),
+}
+
+impl FatalDapError {
+    #[doc(hidden)]
+    pub fn __use_the_macro(s: String) -> Self {
+        FatalDapError(s)
+    }
+}
+
+impl From<VdafError> for DapError {
+    fn from(e: VdafError) -> Self {
+        match e {
+            VdafError::Codec(..) | VdafError::Vdaf(..) => {
+                Self::Transition(TransitionFailure::VdafPrepError)
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct FatalDapError(String);
+
+impl std::error::Error for FatalDapError {}
+
+impl Display for FatalDapError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
```

Comment:
Because the syntax is somewhat weird, can we add a comment here explaining how the macro is used, how the tracing part works, etc.? It would also be useful to document, with examples, all of the variants that the macro accepts.